# How to use the flux_cone class:

First we need to import the `flux_cone` class from flux_class:

In [1]:
from flux_class_vecs import flux_cone


Now we can create our model as an instance of the flux_cone class by initiating it from an sbml file:

In [2]:
model = flux_cone.from_sbml("./Biomodels/bigg/iAB_RBC_283.xml")

Let us first determine the dimension of the reversible metabolic space by calling the class function `get_lin_dim`

In [3]:
model.get_lin_dim()

2

As we can see, the dimension of the reversible metabolic space is 2. This value is now saved as the paramer `lin_dim` of our class instance model:

In [4]:
print(model.lin_dim)

2


To determine the number of minimal proper faces of the flux cone we call the class function `get_mmbs` which determines the minimal metabolic behaviors of the network. The number of MMBs is equal to the number of minimal proper faces.

In [5]:
model.get_mmbs()
print(len(model.mmbs))

587


The fluxcone of the red blood cell has 587 minimal proper faces. Now we can compute the efms in each of the minimal proper faces by calling the function `get_efms_in_mmbs`. This will take a couple of minutes.


In [6]:
model.get_efvs_in_mmbs()
print([len(model.mmb_efvs[i]) for i in range(len(model.mmb_efvs))])

100%|██████████| 587/587 [13:48<00:00,  1.41s/it]


[2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 2, 2, 2, 2, 2, 3, 3, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 

Furthermore we can compute all fully reversible EFMs by calling the function `get_frev_efms`:

In [7]:
model.get_frev_efvs()

[]

In [8]:
sum(len(model.mmb_efvs[i]) for i in range(len(model.mmb_efvs)))

1207

There are a total of 1207 EFMs in the minimal proper faces distributed as seen above and 2 fully reversible EFMs. With this we can compute the total amount of MEMos of the network:

In [9]:
num = 1
for i in [len(model.mmb_efvs[i]) for i in range(len(model.mmb_efvs))]:
    num*=i

In [10]:
print(format(num*2,'e'))


2.694962e+181


In [16]:
import numpy as np
gens = []
for mmb in model.mmb_efvs:
    for efv in mmb:
        gens.append(efv)

In [18]:
print(len(gens))


1207


In [19]:
mem = np.array(gens)

In [20]:
np.shape(mem)

(1207, 469)

In [21]:
np.linalg.matrix_rank(mem)

98

In [22]:
np.shape(model.stoich)

(342, 469)

In [23]:
np.linalg.matrix_rank(model.stoich)

339